api/subscriptions/subscriptions.go

```diff
@@ -31,16 +31,13 @@ type Subscriptions struct {
 	pendingTx      *pendingTx
 	done           chan struct{}
 	wg             sync.WaitGroup
+	logger         log.Logger
```

Comment:
Any reason for moving this?

Reply:
When you call `log.New(...)` it returns the current root with whatever context you pass. If this is declared at a global scope, then the logger will use the discard handler:

`./log/root.go`:
```go
func init() {
	root.Store(&logger{slog.New(DiscardHandler())})
}
```

Reply:
@libotony if you think it's better I can modify the geth from this:
```go
func New(ctx ...interface{}) Logger {
	return Root().With(ctx...)
}
``` 

to this, which allows global initialisation

```go
// New returns a new logger with the given context.
func New(ctx ...interface{}) Logger {
	return &RootWithContext{
		ctx: ctx,
	}
}

// RootWithContext is a logger than can be initialized at a global scope
type RootWithContext struct {
	ctx []interface{}
}

func (r *RootWithContext) Handler() slog.Handler {
	return Root().Handler()
}

func (r *RootWithContext) With(ctx ...interface{}) Logger {
	return Root().With(append(r.ctx, ctx...)...)
}

func (r *RootWithContext) New(ctx ...interface{}) Logger {
	return Root().With(append(r.ctx, ctx...)...)
}

func (r *RootWithContext) Log(level slog.Level, msg string, ctx ...interface{}) {
	Root().Write(level, msg, append(r.ctx, ctx...)...)
}

func (r *RootWithContext) Trace(msg string, ctx ...interface{}) {
	r.Log(LevelTrace, msg, ctx...)
}

func (r *RootWithContext) Debug(msg string, ctx ...interface{}) {
	r.Log(slog.LevelDebug, msg, ctx...)
}

func (r *RootWithContext) Info(msg string, ctx ...interface{}) {
	r.Log(slog.LevelInfo, msg, ctx...)
}

func (r *RootWithContext) Warn(msg string, ctx ...interface{}) {
	r.Log(slog.LevelWarn, msg, ctx...)
}

func (r *RootWithContext) Error(msg string, ctx ...interface{}) {
	r.Log(slog.LevelError, msg, ctx...)
}

func (r *RootWithContext) Crit(msg string, ctx ...interface{}) {
	r.Log(LevelCrit, msg, ctx...)
	os.Exit(1)
}

func (r *RootWithContext) Write(level slog.Level, msg string, ctx ...interface{}) {
	Root().Write(level, msg, append(r.ctx, ctx...)...)
}

func (r *RootWithContext) Enabled(ctx context.Context, level slog.Level) bool {
	return Root().Enabled(ctx, level)
}
```

Reply:
Hmmm, let me do some research about this and get back to you.

Reply:
Looks like it can not be achieved without modifying the log package. Is the current log package a full copy from `ethereum/go-ethereum/log` ?

Reply:
I added this for geth logs:

```go
func SetDefaultGeth(lvl ethlog.Lvl) {
	handler := &EthLogHandler{
		logger: New("pkg", "geth"),
	}
	ethLogHandler := ethlog.NewGlogHandler(handler)
	ethLogHandler.Verbosity(lvl)
	ethlog.Root().SetHandler(ethLogHandler)
}

func (h *EthLogHandler) Log(r *ethlog.Record) error {
	switch r.Lvl {
	case ethlog.LvlCrit:
		h.logger.Crit(r.Msg)
	case ethlog.LvlError:
		h.logger.Error(r.Msg)
	case ethlog.LvlWarn:
		h.logger.Warn(r.Msg)
	case ethlog.LvlInfo:
		h.logger.Warn(r.Msg)
	case ethlog.LvlDebug:
		h.logger.Debug(r.Msg)
	default:
		break
	}

	return nil
}
```

Reply:
> I added this for geth logs. Apart for this its a direct copy
> 
> ```go
> func SetDefaultGeth(lvl ethlog.Lvl) {
> 	handler := &EthLogHandler{
> 		logger: New("pkg", "geth"),
> 	}
> 	ethLogHandler := ethlog.NewGlogHandler(handler)
> 	ethLogHandler.Verbosity(lvl)
> 	ethlog.Root().SetHandler(ethLogHandler)
> }
> 
> func (h *EthLogHandler) Log(r *ethlog.Record) error {
> 	switch r.Lvl {
> 	case ethlog.LvlCrit:
> 		h.logger.Crit(r.Msg)
> 	case ethlog.LvlError:
> 		h.logger.Error(r.Msg)
> 	case ethlog.LvlWarn:
> 		h.logger.Warn(r.Msg)
> 	case ethlog.LvlInfo:
> 		h.logger.Warn(r.Msg)
> 	case ethlog.LvlDebug:
> 		h.logger.Debug(r.Msg)
> 	default:
> 		break
> 	}
> 
> 	return nil
> }
> ```

Can above code move outside of `log` package and then keep the log package the same with upstream? And probably we need a read me like [this one](https://github.com/vechain/thor/blob/master/tracers/readme.md) inside the log package

Reply:
> @libotony if you think it's better I can modify the geth code from this:

Haven't get into the details of this, do you think this one is a good approach or you prefer the eth style?

Reply:
made that change @libotony, but one thing I forgot is that I added license headers to most files there. Should I remove the headers and ignore this directory in the license header check? 

https://github.com/vechain/thor/pull/749/commits/d88e9b9415302b1eb3639b3105c89c9a0bb14463

Reply:
> Haven't get into the details of this, do you think this one is a good approach or you prefer the eth style?

Its annoying that we can't have global loggers. But i guess its just something we need to know going forward if we go ahead with geth implementation. I'm easy whatever we decide there

Reply:
> > Haven't get into the details of this, do you think this one is a good approach or you prefer the eth style?
> 
> Its annoying that we can't have global loggers. But i guess its just something we need to know going forward if we go ahead with geth implementation. I'm easy whatever we decide there

I will take a deeper look around this and come back to you

Reply:
> made that change @libotony, but one thing I forgot is that I added license headers to most files there. Should I remove the headers and ignore this directory in the license header check?
> 
> [d88e9b9](https://github.com/vechain/thor/commit/d88e9b9415302b1eb3639b3105c89c9a0bb14463)

Looks like the license checker allows go-ethereum headers

```
      Copyright(?: \(c\))? \b(19\d{2}|20\d{2})\b The (VeChainThor developers|go-ethereum Authors)
```

Reply:
Ya it does, but the geth files did not come with headers. I added them, so the `log` directly is technically modified.

Reply:
> Ya it does, but the geth files did not come with headers. I added them, so the `log` directly is technically modified.

Ha, license header is good to have :-)